src/serac/numerics/tests/equationsolver.cpp

```diff
@@ -0,0 +1,131 @@
+  mfem::HypreParVector x_computed(&fes);
+
+  std::unique_ptr<mfem::HypreParMatrix> J;
+
+  // Define the types for the test and trial spaces using the function arguments
+  using test_space  = H1<p>;
+  using trial_space = H1<p>;
+
+  // Construct the new functional object using the known test and trial spaces
+  Functional<test_space(trial_space)> residual(&fes, {&fes});
+
+  x_exact.Randomize(0);
+
+  residual.AddDomainIntegral(
+      Dimension<dim>{}, DependsOn<0>{},
+      [&](auto, auto scalar) {
+        auto [u, du_dx] = scalar;
+        auto source     = 0.5 * sin(u);
+        auto flux       = du_dx;
+        return serac::tuple{source, flux};
+      },
+      pmesh);
+
+  StdFunctionOperator residual_opr(
+      fes.TrueVSize(),
+      [&x_exact, &residual](const mfem::Vector& x, mfem::Vector& r) {
+        // TODO this copy is required as the sundials solvers do not allow move assignments because of their memory
+        // tracking strategy
+        // See https://github.com/mfem/mfem/issues/3531
+        const mfem::Vector res = residual(x);
+
+        r = res;
+        r -= residual(x_exact);
+      },
+      [&residual, &J](const mfem::Vector& x) -> mfem::Operator& {
+        auto [val, grad] = residual(differentiate_wrt(x));
+        J                = assemble(grad);
+        return *J;
+      });
+
+  const LinearSolverOptions lin_opts = {.linear_solver  = lin_solver,
+                                        .preconditioner = precond,
+                                        .relative_tol   = 1.0e-10,
+                                        .absolute_tol   = 1.0e-12,
+                                        .max_iterations = 500,
+                                        .print_level    = 1};
+
+  const NonlinearSolverOptions nonlin_opts = {.nonlin_solver  = nonlin_solver,
+                                              .relative_tol   = 1.0e-10,
+                                              .absolute_tol   = 1.0e-12,
+                                              .max_iterations = 100,
+                                              .print_level    = 1};
+
+  EquationSolver eq_solver(nonlin_opts, lin_opts);
+
+  eq_solver.SetOperator(residual_opr);
+
+  eq_solver.Solve(x_computed);
+
+  for (int i = 0; i < x_computed.Size(); ++i) {
+    EXPECT_LT(std::abs((x_computed(i) - x_exact(i))) / x_exact(i), 1.0e-6);
+  }
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    AllEquationSolverTests, EquationSolverSuite,
+    testing::Combine(testing::Values(NonlinearSolver::Newton, NonlinearSolver::LBFGS, NonlinearSolver::KINFullStep,
+                                     NonlinearSolver::KINBacktrackingLineSearch, NonlinearSolver::KINPicard),
+                     testing::Values(LinearSolver::CG, LinearSolver::GMRES, LinearSolver::SuperLU),
+                     testing::Values(Preconditioner::HypreJacobi, Preconditioner::HypreL1Jacobi,
+                                     Preconditioner::HypreGaussSeidel, Preconditioner::HypreAMG)));
```

Comment:
I'm not a gtest superuser-- does this generate a test for each possible combination of {nonlinear, linear, preconditioner} values?

Reply:
Yes - it tests all of the possible combinations.

Reply:
Cool, I look forward to using this sort of pattern in some of our other test suites too!

Reply:
Yep! It's super handy. 
![image](https://user-images.githubusercontent.com/17935880/235745604-e0b3e14b-dc56-48ee-9558-fbb39d79c825.png)